### ISSUE TITLE
Dynamic shape handling for aten.index.Tensor

### ISSUE THREAD

**Issue by qedawkins** [llvm/torch-mlir#1226]
I'm looking to add support for dynamic shapes with multiple indexing tensors for the aten.index.Tensor op. The semantics for the op are summarized as follows.

```
IndexTensor for multiple input tensors broadcasts their shapes to a common
shape and then replaces the indexed dims with the indices given by the
indexing tensors:
x[i_1, i_2, ..., i_M] = result
result[...] = x[i_1[...], i_2[...], ..., i_M[...]]

where the result shape is computed as follows:
1. broadcast i_1, i_2, ..., i_M to a common shape
2. if i_1, i_2, ..., i_M is not contiguous, transpose the broadcasted
   shape to the beginning of the result shape, while removing the
   unchanged dims (marked by None)
3. Otherwise replace the indexed dims with the broadcasted shape

e.g. x shape: [2, 3]
     x[i_1 shape=[4], i_2 shape=[6, 1]] -> Tensor of shape [6, 4]
```

The current approach is to initialize a tensor with the same shape as the output tensor and then lower to a linalg.generic op that iterates over the output shape with affine maps between the index tensors and the dims in the output shape that correspond with the broadcasted shape of the indexing tensors. For example, this e2e test

```
class IndexTensorMultiInputThreeIndexers(torch.nn.Module):

    def __init__(self):
        super().__init__()

    @export
    @annotate_args([
        None,
        ([-1, -1, -1, -1, -1, -1], torch.float32, True),
        ([8, 4, 2], torch.int64, True),
        ([8, 1, 1], torch.int64, True),
        ([4, 2], torch.int64, True),
    ])   
    def forward(self, x, index1, index2, index3):
        return torch.ops.aten.index(x, (None, None, index1, None, index2, index3))


@register_test_case(module_factory=lambda: IndexTensorMultiInputThreeIndexers())
def IndexTensorMultiInputThreeIndexers_basic(module, tu: TestUtils):
    module.forward(tu.rand(1, 2, 4, 4, 5, 3),
            torch.randint(3, (8, 4, 2,)),
            torch.randint(4, (8, 1, 1,)),
            torch.randint(2, (4, 2,)))
```

Lowers to this

```
#map0 = affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1, d2)>
#map1 = affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, 0, 0)>
#map2 = affine_map<(d0, d1, d2, d3, d4, d5) -> (d1, d2)>
#map3 = affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1, d2, d3, d4, d5)>
module attributes {torch.debug_module_name = "IndexTensorMultiInputThreeIndexers"} {
  func.func @forward(%arg0: tensor<?x?x?x?x?x?xf32>, %arg1: tensor<8x4x2xi64>, %arg2: tensor<8x1x1xi64>, %arg3: tensor<4x2xi64>) -> tensor<8x4x2x?x?x?xf32> {
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %c3 = arith.constant 3 : index
    %0 = tensor.dim %arg0, %c0 : tensor<?x?x?x?x?x?xf32>
    %1 = tensor.dim %arg0, %c1 : tensor<?x?x?x?x?x?xf32>
    %2 = tensor.dim %arg0, %c3 : tensor<?x?x?x?x?x?xf32>
    %3 = linalg.init_tensor [8, 4, 2, %0, %1, %2] : tensor<8x4x2x?x?x?xf32>
    %4 = linalg.generic {indexing_maps = [#map0, #map1, #map2, #map3], iterator_types = ["parallel", "parallel", "parallel", "parallel", "parallel", "parallel"]} ins(%arg1, %arg2, %arg3 : tensor<8x4x2xi64>, tensor<8x1x1xi64>, tensor<4x2xi64>) outs(%3 : tensor<8x4x2x?x?x?xf32>) {
    ^bb0(%arg4: i64, %arg5: i64, %arg6: i64, %arg7: f32):
      %5 = linalg.index 3 : index
      %6 = linalg.index 4 : index
      %7 = arith.index_cast %arg4 : i64 to index
      %8 = linalg.index 5 : index
      %9 = arith.index_cast %arg5 : i64 to index
      %10 = arith.index_cast %arg6 : i64 to index
      %11 = tensor.extract %arg0[%5, %6, %7, %8, %9, %10] : tensor<?x?x?x?x?x?xf32>
      linalg.yield %11 : f32
    } -> tensor<8x4x2x?x?x?xf32>
    return %4 : tensor<8x4x2x?x?x?xf32>
  }
}

```
 This approach works well when there is either only one indexing tensor or if the shapes of each index tensor is static. The problem for dynamic shapes comes in when constructing the affine maps used with the linalg.generic op. In particular, the affine mappings for index tensors `a` and `b`, say, are different depending on if `b` has a dim of size 1 that gets broadcasted to some larger size.

For example, in the above lowering we have a tensor of size `(8, 4, 2)` and one of size `(8, 1, 1)` that produces these two affine maps.
```
#map0 = affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1, d2)>
#map1 = affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, 0, 0)>
```

Where if the shapes were instead dynamic we wouldn't be able to construct these affine maps due to not knowing whether two broadcasted dims are equal or if one of them is 1.

One potential solution is to explicitly broadcast each of the indexing tensors to the same shape, but my thinking is that this would be an unnecessary performance hit for this op. My question from here is whether there are any other cases in Torch-MLIR where we handle something like this, otherwise is there something in MLIR that can help?

**Comment by silvasean** [comment]
This should be handleable identical to how we handle broadcasting for elementwise ops. See here:

https://github.com/llvm/torch-mlir/blob/9d6ee48661cefa677950bcd262fe2824a4fa9d17/lib/Conversion/TorchToLinalg/Utils.cpp#L157

In particular, dynamic size-1 broadcast is an error.

**Comment by qedawkins** [comment]
Ah I didn't realize that we didn't support dynamic size-1 broadcast. @vivekkhandelwal1 can you provide a compact test case that shows the case we need for v-diffusion? If we don't need to handle the dynamic size-1 case then this issue can be closed.

**Comment by vivekkhandelwal1** [comment]
> Ah I didn't realize that we didn't support dynamic size-1 broadcast. @vivekkhandelwal1 can you provide a compact test case that shows the case we need for v-diffusion? If we don't need to handle the dynamic size-1 case then this issue can be closed.

```
class IndexTensorMultiInputContiguousOneDimDynamic(torch.nn.Module):

    def __init__(self):
        super().__init__()

    @export
    @annotate_args([
        None,
        ([-1, -1, -1], torch.float32, True),
        ([-1, 1], torch.int64, True),
        ([-1], torch.int64, True),
    ])
    def forward(self, x, index1, index2):
        return torch.ops.aten.index(x, (None, index1, index2,))


@register_test_case(module_factory=lambda: IndexTensorMultiInputContiguousOneDimDynamic())
def IndexTensorMultiInputContiguousOneDimDynamic_basic(module, tu: TestUtils):
    module.forward(tu.rand(5, 4, 3), torch.randint(4, (6, 1)), torch.randint(3, (3,)))
```
